pkg/asset/rhcos/image.go

```diff
@@ -52,101 +55,107 @@ func (i *Image) Dependencies() []asset.Asset {
 func (i *Image) Generate(ctx context.Context, p asset.Parents) error {
 	if oi, ok := os.LookupEnv("OPENSHIFT_INSTALL_OS_IMAGE_OVERRIDE"); ok && oi != "" {
 		logrus.Warn("Found override for OS Image. Please be warned, this is not advised")
-		*i = Image(oi)
+		*i = *MakeAsset(oi)
 		return nil
 	}
 
 	ic := &installconfig.InstallConfig{}
 	p.Get(ic)
 	config := ic.Config
-	osimage, err := osImage(ctx, config)
+	osimageControlPlane, osImageCompute, err := osImage(ctx, config)
 	if err != nil {
 		return err
 	}
-	*i = Image(osimage)
+	*i = Image{osimageControlPlane, osImageCompute}
 	return nil
 }
 
+// For compatibility with before the asset had multi-arch support, returns the same image for both controlPlane and
+// compute.
+//
 //nolint:gocyclo
-func osImage(ctx context.Context, config *types.InstallConfig) (string, error) {
+func osImage(ctx context.Context, config *types.InstallConfig) (string, string, error) {
 	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
 	defer cancel()
 
 	archName := arch.RpmArch(string(config.ControlPlane.Architecture))
 
 	st, err := rhcos.FetchCoreOSBuild(ctx)
 	if err != nil {
-		return "", err
+		return "", "", err
 	}
 	streamArch, err := st.GetArchitecture(archName)
 	if err != nil {
-		return "", err
+		return "", "", err
 	}
 	switch config.Platform.Name() {
 	case aws.Name:
-		if len(config.Platform.AWS.AMIID) > 0 {
-			return config.Platform.AWS.AMIID, nil
+		if ami := config.Platform.AWS.AMIID; len(ami) > 0 {
```

Comment:
we could just call osimage twice in the generate rather than do this - and this is just returning the image matching the cp arch which will not work if the compute arch is different.

Reply:
THB I don't even know why we have `amiID` at the platform level if we can specify it for controlPlane/compute/defaultMachinePlatform. If someone specifies this amiID for nodes with different arches, it's more of user error than anything.

Reply:
maybe it's like an optimization so they don't have to enter it twice :)

Reply:
@patrickdillon Do you know what's the purpose of `amiID` here?

Reply:
> @patrickdillon Do you know what's the purpose of amiID here?

We probably need to do some archaeology to confirm, but I think this was the initial, non-machinepool implementation for custom AMI and was replaced by machinepools so it should be deprecated.

Reply:
Thanks. I'll create a card for that.

Reply:
https://issues.redhat.com/browse/CORS-3608

Reply:
https://github.com/openshift/installer/pull/8755